DuetAI knowledge lookup prompts (#29507)
apache/beam: learning/prompts/documentation-lookup/15_advanced_xlang.md

Prompt:
What is a multi-language pipeline in Apache Beam?
Response:

> Beam lets you combine transforms written in any supported SDK language (currently, [Java](https://beam.apache.org/documentation/programming-guide/#1311-creating-cross-language-java-transforms) and [Python](https://beam.apache.org/documentation/programming-guide/#1312-creating-cross-language-python-transforms)) and use them in one multi-language pipeline. For example, a pipeline that reads from a Python source, processes the data using a Java transform, and writes the data to a Python sink is a multi-language pipeline.

Reviewer: Beam lets you combine transforms written in any supported SDK language (currently, [Java](https://beam.apache.org/documentation/programming-guide/#1311-creating-cross-language-java-transforms) and [Python](https://beam.apache.org/documentation/programming-guide/#1312-creating-cross-language-python-transforms)) and use them in one multi-language pipeline. For example, a pipeline that reads from a Python source processes the data using a Java transform, and writes the data to a Python sink is a multi-language pipeline.

Author: Done

> For example, the [Apache Kafka connector](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/kafka.py) and [SQL transform](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/transforms/sql.py) from the Java SDK can be used in Python pipelines.

Reviewer 2: ```suggestion
For example, you can use the [Apache Kafka connector](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/kafka.py) and [SQL transform](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/transforms/sql.py) from the Java SDK in Python pipelines.
```

> See quickstart examples for [Java](https://beam.apache.org/documentation/sdks/java-multi-language-pipelines) and [Python](https://beam.apache.org/documentation/sdks/python-multi-language-pipelines) to learn how to create a multi-language pipeline.

Reviewer 2: ```suggestion
To learn how to create a multi-language pipeline, see the quickstart examples for [Java](https://beam.apache.org/documentation/sdks/java-multi-language-pipelines) and [Python](https://beam.apache.org/documentation/sdks/python-multi-language-pipelines).
```

> Depending on the SDK language of the pipeline, you can use a high-level SDK-wrapper class, or a low-level transform class to access a cross-language transform. See [Using cross-language transforms](https://beam.apache.org/documentation/programming-guide/#use-x-lang-transforms) section of Apache Beam Documentation.

Reviewer 2: ```suggestion
Depending on the SDK language of the pipeline, you can use a high-level SDK-wrapper class or a low-level transform class to access a cross-language transform. For more information, see [Using cross-language transforms](https://beam.apache.org/documentation/programming-guide/#use-x-lang-transforms).
```

> Developing a cross-language transform involves defining a Uniform Resourse Name(URN) for registering the transform with an expansion service. See [Defining a URN](https://beam.apache.org/documentation/programming-guide/#1314-defining-a-urn) for additional information and examples.

Reviewer 2: ```suggestion
To develop a cross-language transform, you need to define a Uniform Resourse Name(URN) for registering the transform with an expansion service. For more information, see [Defining a URN](https://beam.apache.org/documentation/programming-guide/#1314-defining-a-urn).
```

> Developing a cross-language transform involves defining a Uniform Resourse Name(URN) for registering the transform with an expansion service. See [Defining a URN](https://beam.apache.org/documentation/programming-guide/#1314-defining-a-urn) for additional information. and examples.

Reviewer: Developing a cross-language transform involves defining a Uniform Resourse Name(URN) for registering the transform with an expansion service. See [Defining a URN](https://beam.apache.org/documentation/programming-guide/#1314-defining-a-urn) for additional information and examples.

Author: Done
